### PR TITLE
bump botocore to 1.29.76

### LINKED ISSUE #976: missing client field endpoint_ruleset_resolver in botocore 1.28.0
**Describe the bug**
With botocore 1.28.0 (released 14h ago) aiobotocore 2.4.0 will fail when creating client because botocore 1.28.0 has a new required positional field `endpoint_ruleset_resolver`.
Tested locally on Windows and on Lambda (linux py 3.8)

The same code we have works with botocore 1.27.95
```
  File "C:\Users\xyz\repos\xyz-api\venv2\lib\site-packages\aiobotocore\session.py", line 26, in __aenter__
    self._client = await self._coro
  File "C:\Users\xyz\repos\xyz-api\venv2\lib\site-packages\aiobotocore\session.py", line 193, in _create_client
    client = await client_creator.create_client(
  File "C:\Users\xyz\repos\xyz-api\venv2\lib\site-packages\aiobotocore\client.py", line 70, in create_client
    service_client = cls(**client_args)
TypeError: __init__() missing 1 required positional argument: 'endpoint_ruleset_resolver'
```

**Checklist**
- [ ] I have reproduced in environment where `pip check` passes without errors
- [x] I have provided `pip freeze` results
- [ ] I have provided sample code or detailed way to reproduce
- [x] I have tried the same code in botocore to ensure this is an aiobotocore specific issue
- [ ] I have tried similar code in aiohttp to ensure this is is an aiobotocore specific issue
- [x] I have checked the latest and older versions of aiobotocore/aiohttp/python to see if this is a regression / injection

**pip freeze results**

**Environment:**
 - Python Version: 3.8, 3.9
 - OS name and version: Windows 11, Linux (lambda)

**Additional context**
aioboto3==10.1.0
aiobotocore==2.4.0
aiofiles==0.8.0
aiohttp==3.8.1
aioitertools==0.8.0
aiosignal==1.2.0
altgraph==0.17.2
ansicon==1.89.0
async-timeout==4.0.2
asyncmock==0.4.2
atomicwrites==1.4.0
attrs==21.2.0
aws-cdk.assets==1.137.0
aws-cdk.aws-acmpca==1.137.0
aws-cdk.aws-apigateway==1.137.0
aws-cdk.aws-applicationautoscaling==1.137.0
aws-cdk.aws-athena==1.137.0
aws-cdk.aws-autoscaling==1.137.0
aws-cdk.aws-autoscaling-common==1.137.0
aws-cdk.aws-autoscaling-hooktargets==1.137.0
aws-cdk.aws-certificatemanager==1.137.0
aws-cdk.aws-cloudformation==1.137.0
aws-cdk.aws-cloudfront==1.137.0
aws-cdk.aws-cloudfront-origins==1.137.0
aws-cdk.aws-cloudwatch==1.137.0
aws-cdk.aws-cloudwatch-actions==1.137.0
aws-cdk.aws-codebuild==1.137.0
aws-cdk.aws-codecommit==1.137.0
aws-cdk.aws-codeguruprofiler==1.137.0
aws-cdk.aws-codepipeline==1.137.0
aws-cdk.aws-codestarnotifications==1.137.0
aws-cdk.aws-cognito==1.137.0
aws-cdk.aws-dynamodb==1.137.0
aws-cdk.aws-ec2==1.137.0
aws-cdk.aws-ecr==1.137.0
aws-cdk.aws-ecr-assets==1.137.0
aws-cdk.aws-ecs==1.137.0
aws-cdk.aws-efs==1.137.0
aws-cdk.aws-eks==1.137.0
aws-cdk.aws-elasticloadbalancing==1.137.0
aws-cdk.aws-elasticloadbalancingv2==1.137.0
aws-cdk.aws-events==1.137.0
aws-cdk.aws-events-targets==1.137.0
aws-cdk.aws-globalaccelerator==1.137.0
aws-cdk.aws-glue==1.137.0
aws-cdk.aws-iam==1.137.0
aws-cdk.aws-kinesis==1.137.0
aws-cdk.aws-kinesisfirehose==1.137.0
aws-cdk.aws-kms==1.137.0
aws-cdk.aws-lambda==1.137.0
aws-cdk.aws-lambda-event-sources==1.137.0
aws-cdk.aws-logs==1.137.0
aws-cdk.aws-route53==1.137.0
aws-cdk.aws-route53-targets==1.137.0
aws-cdk.aws-s3==1.137.0
aws-cdk.aws-s3-assets==1.137.0
aws-cdk.aws-s3-notifications==1.137.0
aws-cdk.aws-sam==1.137.0
aws-cdk.aws-secretsmanager==1.137.0
aws-cdk.aws-servicediscovery==1.137.0
aws-cdk.aws-signer==1.137.0
aws-cdk.aws-sns==1.137.0
aws-cdk.aws-sns-subscriptions==1.137.0
aws-cdk.aws-sqs==1.137.0
aws-cdk.aws-ssm==1.137.0
aws-cdk.aws-stepfunctions==1.137.0
aws-cdk.aws-stepfunctions-tasks==1.137.0
aws-cdk.cloud-assembly-schema==1.137.0
aws-cdk.core==1.137.0
aws-cdk.custom-resources==1.137.0
aws-cdk.cx-api==1.137.0
aws-cdk.lambda-layer-awscli==1.137.0
aws-cdk.lambda-layer-kubectl==1.137.0
aws-cdk.lambda-layer-node-proxy-agent==1.137.0
aws-cdk.region-info==1.137.0
aws-xray-sdk==2.10.0
blessed==1.19.0
boto==2.49.0
boto3==1.25.0
botocore==1.28.0
cattrs==1.8.0
certifi==2021.10.8
chalice==1.26.3
chardet==3.0.4
charset-normalizer==2.0.9
click==8.0.3
colorama==0.4.4
constructs==3.3.168
coverage==5.5
demjson==2.2.4
dirtyjson==1.0.7
ecdsa==0.17.0
frozenlist==1.2.0
future==0.18.2
idna==2.10
iniconfig==1.1.1
inquirer==2.8.0
jinxed==1.1.0
jmespath==0.10.0
jsii==1.56.0
mock==4.0.3
multidict==5.2.0
mypy-extensions==0.4.3
packaging==21.3
pefile==2021.9.3
Pillow==8.2.0
pluggy==1.0.0
psutil==5.8.0
publication==0.0.3
py==1.11.0
pyasn1==0.4.8
pyinstaller==4.10
pyinstaller-hooks-contrib==2022.2
pyparsing==3.0.7
pytest==6.2.5
pytest-asyncio==0.17.2
pytest-cov==3.0.0
pytest-html==3.1.1
pytest-metadata==2.0.1
python-dateutil==2.8.2
python-editor==1.0.4
python-jose==3.3.0
pywin32-ctypes==0.2.0
PyYAML==6.0
qrcode==7.3.1
readchar==2.0.1
requests==2.25.1
rsa==4.8
s3transfer==0.6.0
six==1.16.0
toml==0.10.2
typing-extensions==3.10.0.2
unittest-xml-reporting==3.0.4
urllib3==1.25.11
watchtower==3.0.0
wcwidth==0.2.5
wrapt==1.13.3
yarl==1.7.2

### LINKED ISSUE #993: Bump botocore
started work in https://github.com/aio-libs/aiobotocore/pull/974 however needs research into new async codepaths

### LINKED ISSUE #995: Recent AWS SSO Config change are raising InvalidConfigError on old botocore
**Describe the bug**
After an update on AWS CLI and reconfiguring my SSO session, I've started getting `InvalidConfigError` on projects using aiobotocore because of the botocore version. The exception is as follows:

```
/tmp/test-sso/test-sso.py:9: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
Traceback (most recent call last):
  File "/tmp/test-sso/test-sso.py", line 10, in <module>
    loop.run_until_complete(main())
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/tmp/test-sso/test-sso.py", line 6, in main
    cred = await session.get_credentials()
  File "/tmp/test-sso/venv/lib/python3.10/site-packages/aiobotocore/session.py", line 79, in get_credentials
    self._credentials = await (
  File "/tmp/test-sso/venv/lib/python3.10/site-packages/aiobotocore/credentials.py", line 951, in load_credentials
    creds = await provider.load()
  File "/tmp/test-sso/venv/lib/python3.10/site-packages/aiobotocore/credentials.py", line 1037, in load
    sso_config = self._load_sso_config()
  File "/tmp/test-sso/venv/lib/python3.10/site-packages/botocore/credentials.py", line 2236, in _load_sso_config
    raise InvalidConfigError(
botocore.exceptions.InvalidConfigError: The profile "mlengineer" is configured to use SSO but is missing required configuration: sso_start_url, sso_region
```

This is very inconvenient because now I have to export my AWS environment variables every time I want to deal with anything related to aiobotocore.

**Checklist**
- [x] I have reproduced in environment where `pip check` passes without errors
- [x] I have provided `pip freeze` results
- [x] I have provided sample code or detailed way to reproduce
- [ ] I have tried the same code in botocore to ensure this is an aiobotocore specific issue
- [ ] I have tried similar code in aiohttp to ensure this is is an aiobotocore specific issue
- [x] I have checked the latest and older versions of aiobotocore/aiohttp/python to see if this is a regression / injection

**pip freeze results**

```
./venv/bin/python -m pip freeze                                                                         ian@batata
aiobotocore==2.4.2
aiohttp==3.8.3
aioitertools==0.11.0
aiosignal==1.3.1
async-timeout==4.0.2
attrs==22.2.0
botocore==1.27.59
charset-normalizer==2.1.1
frozenlist==1.3.3
idna==3.4
jmespath==1.0.1
multidict==6.0.4
python-dateutil==2.8.2
six==1.16.0
urllib3==1.26.14
wrapt==1.14.1
yarl==1.8.2
```

**Environment:**
 - Python Version: 3.10.9
 - OS name and version: Arch Linux

**Sample code**

```python
import asyncio
from aiobotocore.session import get_session

async def main():
    session = get_session()
    cred = await session.get_credentials()
    print(cred)

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

**Additional context**

You must first have a working SSO config with `aws configure sso` and `aws sso login` with the latest AWS CLI. This will create an invalid configuration for aiobotocore.

The exception is happening in the `botocore` library itself. I've bisected the botocore version to find where it starts working, and I can state that `botocore==1.29.10` is the first version that starts working with the new SSO config scheme.

### LINKED ISSUE #996: Project dependencies.
`aiobotocore 2.4.2 has requirement botocore<1.27.60,>=1.27.59, but you have botocore 1.29.68.` So your just making pointless dependency lock? When not remove the dependency and integrate it?

### PR DESCRIPTION
**Pull request by jakob-keller**
### Description of Change
The botocore dependency has not been updated in a while, causing a number of downstream issues and incompatibilities.

### Assumptions
Diff between previous and new botocore sources: https://github.com/boto/botocore/compare/1.27.59...1.29.76

### Checklist for All Submissions
* [x] I have added change info to [CHANGES.rst](https://github.com/aio-libs/aiobotocore/blob/master/CHANGES.rst)
* [x] If this is resolving an issue (needed so future developers can determine if change is still necessary and under what conditions) (can be provided via link to issue with these details): 
  * [x] Detailed description of issue closes #976, closes #993, closes #995, and closes #996, 
  * [ ] Alternative methods considered (if any)
  * [x] How issue is being resolved: follow instructions in [CONTRIBUTING.rst](https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst#how-to-upgrade-botocore)
  * [ ] How issue can be reproduced

### Checklist when updating botocore and/or aiohttp versions

* [x] I have read and followed [CONTRIBUTING.rst](https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst#how-to-upgrade-botocore)
* [x] I have updated test_patches.py where/if appropriate (also check if no changes necessary)
* [x] I have ensured that the awscli/boto3 versions match the updated botocore version

### CODE REVIEW

**Comment by thehesiod** [comment]
see work I started in https://github.com/aio-libs/aiobotocore/pull/974, we need to research new codepath: SSOTokenProvider._refresher

**Comment by thehesiod** [comment]
oh, I see something along those lines, does this PR account for that?

**Comment by jakob-keller** [comment]
> oh, I see something along those lines, does this PR account for that?

Not explicitly. I did not encounter anything that looked like new blocking codepaths to me. I would appreciate a thorough review or two, since I am new to this codebase.

**Comment by jakob-keller** [comment]
> oh, I see something along those lines, does this PR account for that?

I just pushed a commit that supports async token refresh. Please note that this is currently untested. Is anyone able to give it a try and/or suggest unit tests?

**Comment by thehesiod** [comment]
aha, @jakob-keller you da man.  Ok for stuff like this I typically pulled tests from [botocore](https://github.com/boto/botocore/tree/develop/tests).  You can see some of our adaptations here: https://github.com/aio-libs/aiobotocore/tree/master/tests/boto_tests. where possible keeping the tests as similar as possible

**Comment by jakob-keller** [comment]
> aha, @jakob-keller you da man. Ok for stuff like this I typically pulled tests from [botocore](https://github.com/boto/botocore/tree/develop/tests). You can see some of our adaptations here: https://github.com/aio-libs/aiobotocore/tree/master/tests/boto_tests. where possible keeping the tests as similar as possible

OK, I will give it a shot.

I just discovered another new blocking codepath in regions.py. Will address that, too.

**Comment by thehesiod** [comment]
oops, forgot probably easier doing it in https://github.com/aio-libs/aiobotocore/tree/master/tests/python3.8/boto_tests, no reason to make it 3.6 compatible

**Comment by jakob-keller** [comment]
OK, I addressed the failing unit tests and the second blocking codepath. It's working locally on my machine. Could you trigger CI again, please?

I haven't gotten around to adding new tests yet. Will try to do that in the coming days.

**Comment by jakob-keller** [comment]
@thehesiod Could you trigger CI again, please? I want to make sure existing tests are fine before I add anything new. Thanks!

**Comment by thehesiod** [comment]
@jakob-keller thanks again, sorry for delay, work + fam have been non-stop.

**Comment by jakob-keller** [comment]
OK, I had a look at the unit tests from upstream and added and/or modified those that I believe are relevant.

@thehesiod: Could you please run CI again to check if there are any remaining issues?

**Comment by jakob-keller** [comment]
Well it looks like test coverage is subpar. Is that a hard requirement? I am not terribly familiar with the test suits of this repo nor upstream, which makes me rather doubtful that I can contribute much more without specific guidance. Any input? Thanks!

**Comment by thehesiod** [comment]
I think the tests may not be running because they're not marked moto: https://github.com/aio-libs/aiobotocore/blob/master/tests/test_monitor.py#L6.  Right now we only run moto tests: https://github.com/aio-libs/aiobotocore/blob/master/Makefile#L25 you can run the tests locally as well to check

**Comment by tibbe** [comment]
I would love to see this. I'm trying to decide whether to use aioboto3 in my project and it being ~7 months behind boto3 worries me a bit. For example, if there's a critical bug fix to boto3 and it takes long for it to be picked up by aioboto3 that would be a cause for concern.

I of course understand this is open source and people have limited time. Just wanted to share my thoughts of adopting this package as a potential user.

**Comment by jakob-keller** [comment]
> I think the tests may not be running because they're not marked moto: https://github.com/aio-libs/aiobotocore/blob/master/tests/test_monitor.py#L6. Right now we only run moto tests: https://github.com/aio-libs/aiobotocore/blob/master/Makefile#L25 you can run the tests locally as well to check

OK, I believe it's ready for another spin. Awaiting CI run...

**Comment by jakob-keller** [comment]
Well, this time, I believe we are pretty close. Anything else I can do to move this forward?

**Comment by jakob-keller** [comment]
@thehesiod What do you think about the state of this PR? Anything else that I can contribute or need to address?

**Comment by thehesiod** [comment]
@jakob-keller I think it's great, just need a sec to look over it

**Comment by thehesiod** [comment]
looking now

File: aiobotocore/tokens.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+        if self._is_expired():
+            # Fresh credentials should never be expired
+            raise TokenRetrievalError(
+                provider=self.method,
+                error_msg="Token has expired and refresh failed",
+            )
+
+
+class AioSSOTokenProvider(SSOTokenProvider):
```
Comment (thehesiod) [inline aiobotocore/tokens.py:84]: I'd prefer this being in credentials.py like in botocore but no biggie can move later

Comment (thehesiod) [reply aiobotocore/tokens.py:84]: that will make later diffing easier

Comment (thehesiod) [reply aiobotocore/tokens.py:84]: nm deleting thread, i was confused

File: aiobotocore/session.py [hunk 2]
```diff
@@ -167,6 +171,7 @@ async def _create_client(
             )
         else:
             credentials = await self.get_credentials()
+        auth_token = self.get_auth_token()
```
Comment (thehesiod) [inline aiobotocore/session.py:174]: nice work piping this through!

**Review by thehesiod — APPROVED** [review]

**Comment by thehesiod** [comment]
looks great! Thanks so much for all your hard work! This was one of the more complicated merges.  I'll get this out today

**Comment by thehesiod** [comment]
we should point to this PR why bumping is not so simple :)

**Comment by thehesiod** [comment]
> I would love to see this. I'm trying to decide whether to use aioboto3 in my project and it being ~7 months behind boto3 worries me a bit. For example, if there's a critical bug fix to boto3 and it takes long for it to be picked up by aioboto3 that would be a cause for concern.
> 
> I of course understand this is open source and people have limited time. Just wanted to share my thoughts of adopting this package as a potential user.

ya I have limited time now with kids and start-up life, have any other volunteers for help with getting releases out sooner?  This one took longer than expected due to the complications you can see from this PR.  They added several new codepaths that needed to be async'ified

**Comment by jakob-keller** [comment]
> ya I have limited time now with kids and start-up life

I feel you!

> have any other volunteers for help with getting releases out sooner?

Count me in

**Comment by paked** [comment]
Really appreciate the work here guys. Thanks so much!

**Comment by martindurant** [comment]
Is a release planned with this or more recent botocore dependency? s3fs is being hit by the large number of releases botocore has had since the last 2.5.0 https://github.com/dask/dask/issues/10276

**Comment by jakob-keller** [comment]
> Is a release planned with this or more recent botocore dependency?

This has been released as `aiobotocore 2.5.0` on Mar 7, 2023. Which minimum release of `botocore` would you require?

**Pull request merged**